File: userland/apps/asemics/commands.c
#include "asemics.h"
#include <string.h>

int parseCommands(char *cmd, int len) {
  int cmds = 0;
  if (len == 1) {
    return 0;
  }

  int i = 1;

  while (cmd[i] != '\0') {
    switch (cmd[i]) {
    case 'w':
      cmds |= SAVE;
      break;
    case 'q':
      cmds |= QUIT;
      break;
    case '!':
      cmds |= FORCE;
      break;
    default:
      return -1;
      break;
    }
    i++;
  }
  return cmds;
}

void execCommands() {
  char *cmd = E.ui.cmdline;
  int len = strlen(cmd);

  int cmds = parseCommands(cmd, len);
  bool force = false;

  if (cmds == -1) {
    // TODO: unknown cmd
  }

  if (cmds == 0) {
    // TODO: no-op
  }

  if (cmds & SAVE) {
    editorSave();
  }
  if (cmds & QUIT) {
    if (cmds & FORCE) {
      force = true;
      editorQuit(force);
    } else {
      editorQuit(force);
    }
  }

  E.mode = MODE_NORMAL;
}

void commandInsertChar(int c) {
  int n = strlen(E.ui.cmdline);
  if (n >= 80) {
    return;
  }
  E.ui.cmdline[n] = c;
  E.ui.cmdline[n + 1] = '\0';
  return;
}

void commandDelChar(void) {
  int n = strlen(E.ui.cmdline);
  if (n == 1) return;
  E.ui.cmdline[n - 1] = '\0';
  return;
}

void editorDrawCmdline(WriteBuf *wb) {
  writeBufAppend(wb, "\x1b[K", 3);
  int cmdlen = strlen(E.ui.cmdline);
  if (cmdlen > E.viewport.width) cmdlen = E.viewport.width;
  writeBufAppend(wb, E.ui.cmdline, cmdlen);
}

File: userland/apps/asemics-rs/src/commands.rs
use crate::consts::{FORCE, QUIT, SAVE, SCREEN_CLEAR_LINE};
use crate::editor::{Editor, EditorMode};

pub fn parse_commands(cmd: &[u8]) -> i32 {
    let mut cmds = 0;
    if cmd.len() == 1 {
        return 0;
    }

    for &b in cmd.get(1..).unwrap_or(&[]) {
        match b {
            b'w' => cmds |= SAVE,
            b'q' => cmds |= QUIT,
            b'!' => cmds |= FORCE,
            _ => return -1,
        }
    }
    cmds
}

impl Editor {
    pub fn exec_commands(&mut self) {
        let cmds = parse_commands(&self.ui.cmdline);

        /* note: an unknown command parses to -1, which has every bit set, so it
         * behaves as save + force-quit — same as the C version */

        if cmds & SAVE != 0 {
            self.save();
        }
        if cmds & QUIT != 0 {
            self.quit(cmds & FORCE != 0);
        }

        self.mode = EditorMode::Normal;
    }

    pub fn command_insert_char(&mut self, c: i32) {
        let n = self.ui.cmdline.len();
        if n >= 80 {
            return;
        }
        self.ui.cmdline.push(c as u8);
    }

    pub fn command_del_char(&mut self) {
        let n = self.ui.cmdline.len();
        if n == 1 {
            return;
        }
        self.ui.cmdline.pop();
    }

    pub fn draw_cmdline(&self, wb: &mut Vec<u8>) {
        wb.extend_from_slice(SCREEN_CLEAR_LINE);
        let mut cmdlen = self.ui.cmdline.len() as i32;
        if cmdlen > self.viewport.width {
            cmdlen = self.viewport.width;
        }
        wb.extend_from_slice(&self.ui.cmdline[..cmdlen as usize]);
    }
}
